src/core/connection.c

```diff
@@ -6099,6 +6186,201 @@ QuicConnUpdatePeerPacketTolerance(
+    UdpConfig.InterfaceIndex = 0;
+#ifdef QUIC_COMPARTMENT_ID
+    UdpConfig.CompartmentId = Connection->Configuration->CompartmentId;
+#endif
+#ifdef QUIC_OWNING_PROCESS
+    UdpConfig.OwningProcess = Connection->Configuration->OwningProcess;
+#endif
+
+    QUIC_BINDING* NewBinding = NULL;
+    Status =
+        QuicLibraryGetBinding(
+            &UdpConfig,
+            &NewBinding);
+    if (QUIC_FAILED(Status)) {
+        return Status;
+    }
+
+    if (!Connection->State.ShareBinding) {
+        QUIC_CID_SLIST_ENTRY* SourceCid = QuicCidNewNullSource(Connection);
+        if (SourceCid == NULL) {
+            QuicLibraryReleaseBinding(NewBinding);
+            return QUIC_STATUS_OUT_OF_MEMORY;
+        }
+
+        Connection->NextSourceCidSequenceNumber++;
+        QuicTraceEvent(
+            ConnSourceCidAdded,
+            "[conn][%p] (SeqNum=%llu) New Source CID: %!CID!",
+            Connection,
+            SourceCid->CID.SequenceNumber,
+            CASTED_CLOG_BYTEARRAY(SourceCid->CID.Length, SourceCid->CID.Data));
+        CxPlatListPushEntry(&Connection->SourceCids, &SourceCid->Link);
+
+        if (!QuicBindingAddSourceConnectionID(NewBinding, SourceCid)) {
+            QuicLibraryReleaseBinding(NewBinding);
+            return QUIC_STATUS_OUT_OF_MEMORY;
+        }
+    } else {
+        if (!QuicBindingAddAllSourceConnectionIDs(NewBinding, Connection)) {
+            QuicLibraryReleaseBinding(NewBinding);
+            return QUIC_STATUS_OUT_OF_MEMORY;
+        }
+    }
+
+    if (Connection->PathsCount > 1) {
+        //
+        // Make room for the new path (at index 1).
+        //
+        CxPlatMoveMemory(
+            &Connection->Paths[2],
+            &Connection->Paths[1],
+            (Connection->PathsCount - 1) * sizeof(QUIC_PATH));
+    }
+
+    QUIC_PATH* Path = &Connection->Paths[1];
+    QuicPathInitialize(Connection, Path);
+    Connection->PathsCount++;
+
+    Path->Binding = NewBinding;
+
+    QuicBindingGetLocalAddress(
+        Path->Binding,
+        &Path->Route.LocalAddress);
+    CxPlatCopyMemory(&Path->Route.RemoteAddress, &Connection->Paths[0].Route.RemoteAddress, sizeof(QUIC_ADDR));
+
+    Path->Allowance = UINT32_MAX;
+
+    Path->DestCid = NewDestCid;
+    QUIC_CID_SET_PATH(Connection, Path->DestCid, Path);
+    Path->DestCid->CID.UsedLocally = TRUE;
+
+    CXPLAT_DBG_ASSERT(Path->DestCid != NULL);
+    QuicPathValidate(Path);
+    Path->SendChallenge = TRUE;
+    Path->PathValidationStartTime = CxPlatTimeUs64();
+
+    CxPlatRandom(sizeof(Path->Challenge), Path->Challenge);
+
+    QuicTraceEvent(
+        ConnLocalAddrAdded,
+        "[conn][%p] New Local IP: %!ADDR!",
+        Connection,
+        CASTED_CLOG_BYTEARRAY(sizeof(Path->Route.LocalAddress), &Path->Route.LocalAddress));
+
+    QuicSendSetSendFlag(&Connection->Send, QUIC_CONN_SEND_FLAG_PATH_CHALLENGE);
+
+    return QUIC_STATUS_SUCCESS;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+static
+QUIC_STATUS
+QuicConnRemoveLocalAddress(
```

Comment:
There doesn't appear to be any test coverage for the Remove address function here. Could you add a test for that?